## 46-881 Class 5

### Python pandas: Data Exploration and Wrangling (Part 1)

Amr Farahat

CMU / Tepper

2022-11-25

---

## Python pandas: Importing data

In [1]:
import numpy as np
import pandas as pd

#### Introducing the PIT departures dataset

Place the accompanying dataset "PIT_2021.csv" **in the same folder location as this Notebook**. The dataset has flight delay information pertainig to all departures out of PIT (Pittsburgh International Airport) in 2021. 

Data source: Bureau of Transportation Statistics.
Covers domestic flights only
https://www.transtats.bts.gov/DL_SelectFields.aspx?gnoyr_VQ=FGJ&QO_fu146_anzr=b0-gvzr

The dataset contains scheduled and actual departure and arrival times reported by certified U.S. air carriers that account for at least one percent of domestic scheduled passenger revenues. The data is collected by the Office of Airline Information, Bureau of Transportation Statistics (BTS).

Accompanying reference files "Documentation.xlsx" and "Carrier_codes.xlsx" are available on Canvas.

#### Importing csv data as a DataFrame

In [2]:
df = pd.read_csv("PIT_2021.csv")

#### Quick inspection

In [3]:
%whos

Variable   Type         Data/Info
---------------------------------
df         DataFrame           YEAR  MONTH  DAY_O<...>[32620 rows x 33 columns]
np         module       <module 'numpy' from '/op<...>kages/numpy/__init__.py'>
pd         module       <module 'pandas' from '/o<...>ages/pandas/__init__.py'>


To find the dimensions of df

In [4]:
df.shape

(32620, 33)

*We have 32620 fights, described by 33 columns, flown out of PIT in 2021. These are all the domestic flights flown out of PIT by certified U.S. carriers with at least 1% of domestic scheduled passenger revenue.*

In [5]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32620 entries, 0 to 32619
Data columns (total 33 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   YEAR                 32620 non-null  int64  
 1   MONTH                32620 non-null  int64  
 2   DAY_OF_MONTH         32620 non-null  int64  
 3   DAY_OF_WEEK          32620 non-null  int64  
 4   FL_DATE              32620 non-null  object 
 5   OP_CARRIER           32620 non-null  object 
 6   OP_CARRIER_FL_NUM    32620 non-null  int64  
 7   ORIGIN               32620 non-null  object 
 8   ORIGIN_CITY_NAME     32620 non-null  object 
 9   ORIGIN_STATE_ABR     32620 non-null  object 
 10  DEST                 32620 non-null  object 
 11  DEST_CITY_NAME       32620 non-null  object 
 12  DEST_STATE_ABR       32620 non-null  object 
 13  CRS_DEP_TIME         32620 non-null  int64  
 14  DEP_TIME             32177 non-null  float64
 15  DEP_DELAY            32177 non-null 

#### Viewing the DataFrame 

Jupyter notebooks do not have the nice viewer functionality of RStudio :(. There are specialized packages or add-ins you can use for that but we won't cover them here.

The code cell below to disable limits on the number of columns displayed in Jupyter (so you can scroll across and inspect all columns)  

In [6]:
pd.options.display.max_columns = None

Warning: Be careful not to set max_rows to None - there's enough space or memory to display the entire data frame at once. 

In [7]:
df

,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
0,2021,1,10,7,1/10/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1900.0,-1.0,0.0,0.0,1900-1959,2025,2023.0,-2.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
1,2021,1,4,1,1/4/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,756.0,-4.0,0.0,0.0,0800-0859,921,904.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
2,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,754.0,-6.0,0.0,0.0,0800-0859,921,903.0,-18.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
3,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3989,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1245,1243.0,-2.0,0.0,0.0,1200-1259,1403,1339.0,-24.0,0.0,0.0,1400-1459,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
4,2021,1,11,1,1/11/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,807.0,7.0,7.0,0.0,0800-0859,930,913.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32615,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2407,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,1639,1645.0,6.0,6.0,0.0,1600-1659,1859,1853.0,-6.0,0.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32616,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2109,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,838,833.0,-5.0,0.0,0.0,0800-0859,1022,1019.0,-3.0,0.0,0.0,1000-1059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32617,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2108,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,610,604.0,-6.0,0.0,0.0,0600-0659,835,822.0,-13.0,0.0,0.0,0800-0859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32618,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1810,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,1600,1704.0,64.0,64.0,1.0,1600-1659,1744,1841.0,57.0,57.0,1.0,1700-1759,0.0,NaN,0.0,57.0,0.0,0.0,0.0,0.0


Inspect the first few rows and last few rows of df. Note the default index column 0 .. length-1.

In [8]:
df.head(8)

,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
0,2021,1,10,7,1/10/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1900.0,-1.0,0.0,0.0,1900-1959,2025,2023.0,-2.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
1,2021,1,4,1,1/4/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,756.0,-4.0,0.0,0.0,0800-0859,921,904.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
2,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,754.0,-6.0,0.0,0.0,0800-0859,921,903.0,-18.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
3,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3989,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1245,1243.0,-2.0,0.0,0.0,1200-1259,1403,1339.0,-24.0,0.0,0.0,1400-1459,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
4,2021,1,11,1,1/11/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,807.0,7.0,7.0,0.0,0800-0859,930,913.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
5,2021,1,22,5,1/22/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1848.0,-13.0,0.0,0.0,1900-1959,2025,2019.0,-6.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
6,2021,1,22,5,1/22/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,759.0,-1.0,0.0,0.0,0800-0859,930,923.0,-7.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
7,2021,1,5,2,1/5/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1853.0,-8.0,0.0,0.0,1900-1959,2025,2001.0,-24.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN


In [9]:
df.tail()

,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
32615,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2407,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,1639,1645.0,6.0,6.0,0.0,1600-1659,1859,1853.0,-6.0,0.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32616,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2109,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,838,833.0,-5.0,0.0,0.0,0800-0859,1022,1019.0,-3.0,0.0,0.0,1000-1059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32617,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2108,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,610,604.0,-6.0,0.0,0.0,0600-0659,835,822.0,-13.0,0.0,0.0,0800-0859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32618,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1810,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,1600,1704.0,64.0,64.0,1.0,1600-1659,1744,1841.0,57.0,57.0,1.0,1700-1759,0.0,NaN,0.0,57.0,0.0,0.0,0.0,0.0
32619,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1226,PIT,"Pittsburgh, PA",PA,EWR,"Newark, NJ",NJ,1635,1630.0,-5.0,0.0,0.0,1600-1659,1802,1812.0,10.0,10.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN


#### read_csv options

In [10]:
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]', sep=<no_default>, delimiter=None, header='infer', names=<no_default>, index_col=None, usecols=None, squeeze=None, prefix=<no_default>, mangle_dupe_cols=True, dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=None, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None, compression: 'CompressionOptions' = 'infer', thousands=None, decimal: 'str' = '.', lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None, comment=None, encoding=None, encoding_errors: 'str | None' = 'strict', dialect=None, error_bad_li

#### Reading other data formats

- read_table (general delimited data)
- read_fwf
- read_excel
- read_json
- read_xml
- read_html

#### Writing data to text format

In [11]:
df_first5rows = df.head()  

In [12]:
df_first5rows.to_csv("sample.csv")

---
## END